### PR TITLE
feat(Validium): refactor batch commit data generation

### PR DESCRIPTION
**Pull request by Oppen**
## What ❔

Reworks `L1BatchCommitDataGenerator`, `CommitBatches` and `CommitBatchInfo`.
This PR moves most of the encoding responsibility to per-operation-mode
structures derived from `CommitBatches` and `CommitBatchInfo`, modifying
their `Tokenize` implementations. In turn, the `L1BatchCommitDataGenerator`
implementations now differ only on which variant they instantiate to
delegate the request for tokenization.
After that, all aggregation logic and data generators is moved to `eth_sender`.

The goal is to apply [these recommendations](https://github.com/matter-labs/zksync-era/pull/1015#discussion_r1489027643).

NOTE: the following tests are failing both here and in the base branch:
```
     Summary [ 230.521s] 853 tests run: 850 passed, 3 failed, 2 skipped
        FAIL [   1.969s] zksync_core eth_sender::tests::correct_order_for_confirmations
        FAIL [   1.883s] zksync_core eth_sender::tests::skipped_l1_batch_at_the_start
        FAIL [   1.848s] zksync_core eth_sender::tests::skipped_l1_batch_in_the_middle
```

## Why ❔

The original changes for the L1 publish behaviour were a bit disperse over the
codebase, leading to hard to understand code and a more complex dependency graph.

## Checklist

- [ ] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.
- [ ] Code has been formatted via `zk fmt` and `zk lint`.
- [ ] Spellcheck has been run via `zk spellcheck`.
- [ ] Linkcheck has been run via `zk linkcheck`.

### CODE REVIEW

**Comment by Oppen** [comment]
Converted to draft because I missed some changes. Some places were left hardcoded to `CommitBatchesRollup` as a stop-gap and never changed.

File: core/lib/l1_contract_interface/src/i_executor/methods/commit_batches.rs [hunk 1]
```diff
@@ -1,33 +1,44 @@
...
-};
+use zksync_types::{commitment::L1BatchWithMetadata, ethabi::Token};
 
 use crate::{
-    i_executor::structures::{CommitBatchInfo, StoredBatchInfo},
+    i_executor::structures::{CommitBatchInfoRollup, CommitBatchInfoValidium, StoredBatchInfo},
     Tokenizable, Tokenize,
 };
 
 /// Input required to encode `commitBatches` call.
```
Comment (popzxc) [inline core/lib/l1_contract_interface/src/i_executor/methods/commit_batches.rs:8]: To add clarity: "...for a contract running in a Rollup mode". And similar additions can be made in other places.

File: core/lib/l1_contract_interface/src/i_executor/structures/commit_batch_info.rs [hunk 2]
```diff
@@ -88,3 +121,55 @@ fn pre_boojum_into_token(l1_batch_commit_with_metadata: &L1BatchWithMetadata) ->
         ),
     ])
 }
+
+fn encode_l1_commit(l1_batch_with_metadata: &L1BatchWithMetadata, pubdata: Token) -> Vec<Token> {
```
Comment (popzxc) [inline core/lib/l1_contract_interface/src/i_executor/structures/commit_batch_info.rs:125]: Looks like you're only using `header` and `metadata` fields in this function. So, you can slightly rework this function:

```rust
fn encode_l1_commit(header: &L1BatchHeader, metadata: &L1BatchMetadata, pubdata_input: Option<&L1BatchWithMetadata>) -> Vec<Token>
```

This way you won't need separate `validium_mode_l1_commit_data` and `rollup_mode_l1_commit_data`

File: core/lib/zksync_core/src/eth_sender/l1_batch_commit_data_generator.rs [hunk 3]
```diff
@@ -0,0 +1,69 @@
+use zksync_l1_contract_interface::{
+    i_executor::{
+        methods::{CommitBatchesRollup, CommitBatchesValidium},
+        structures::{CommitBatchInfoRollup, CommitBatchInfoValidium},
+    },
+    Tokenizable, Tokenize,
+};
+use zksync_types::{commitment::L1BatchWithMetadata, ethabi::Token};
+
+pub trait L1BatchCommitDataGenerator
```
Comment (popzxc) [inline core/lib/zksync_core/src/eth_sender/l1_batch_commit_data_generator.rs:10]: A doc-comment here would be appreciated.

**Review by popzxc — COMMENTED** [review]
Nice progress here!

File: core/lib/l1_contract_interface/src/i_executor/methods/commit_batches.rs (hunk 1 above)
Comment (Oppen) [reply core/lib/l1_contract_interface/src/i_executor/methods/commit_batches.rs:8]: Good call.

**Review by popzxc — APPROVED** [review]

**Pull request merged**